### PR TITLE
require> enhancement

### LINKED ISSUE #712: require operator does not honor past session's success/failure when retrying
I am using [require operator](http://docs.digdag.io/operators/require.html) to check the status of the previous session of the same workflow. The workflow should -
1. continue if the last session was successful,
2. fail if the last session failed,
3. start last session if the last session does not exist, and,
4. wait for the last session's completion if it's pending.

The above four conditions work perfectly when the session is started by the scheduler or is started using `digdag start` command without the `--retry` option. 

But when I retry a session (either successful or failed) using `digdag retry` or `digdag start --retry` or `digdag backfill`, the require operator starts the last session (condition 3 above) regardless of the status of the last session. In other words, last session is started again even if its status is successful or failed.

### PR DESCRIPTION
**Pull request by yoyama**
This PR is follow up PR to #1405 and enhance require> operator.

- Introduce rerun_on: (none,failed,all).
- Ignore retry_attempt_name from out of operator. (fix #712 )
- exponential backoff for task retry (max 10 secs)

To exclude #1405 fix from file changes, following view may be good.
https://github.com/treasure-data/digdag/pull/1409/files/48b9ae47b4554f8b5b1fa0fadcc1a28842cd07b7..ef1f5fdeb1b8f248d281b5fb989f714b5b75649e

Following is the flow chart of require>
WM_815_Google_Slides

### CODE REVIEW

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 1]
```diff
@@ -65,42 +68,157 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
                         retryAttemptName,
-                        overrideParams);
+                        overrideParams));
+            }
+            catch (SessionAttemptConflictException ex) {
+                sessionAttemptConflictException = Optional.of(ex);
+            }
+            catch (ResourceNotFoundException ex) {
+                throw new TaskExecutionException(String.format(ENGLISH, "Dependent workflow does not exist. %s, workflowName:%s",
+                        projectIdentifier.transform((p)->p.toString()).or(""), workflowName ));
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:125]: There is an unexpected space between `workflowName` and `)`.

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 2]
```diff
@@ -64,46 +68,109 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
+             *  First of all, try to start attempt by startSession()
+             *  If no attempt exists (no conflict), it return new StoredSessionAttempt.
+             *    - set state param "require_kicked" to true.
+             *    - task is retried to wait for done by nextPolling.
+             *  If something errors happen, it will throw following exceptions.
+             *    - ResourceNotFoundException ... workflow ,project name(or id) are wrong. -> processed as deterministic error
+             *    - ResourceLimitExceededException ... this exception should be deterministic error
+             *    - SessionAttemptConflictException ... this is not error.
+             *      - If the conflict attempt is still running, wait for until done by nextPolling.
+             *      - If done, check the state param "require_kicked" and whether the attempt is kicked by this require> or not.
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:100]: What about the case the conflict attempt failed like `If failed, xxx`?

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 1 above)
Comment (komamitsu) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:125]: [minor] `(p)->p.toString()` can be replaced with `ProjectIdentifier::toString`

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 3]
```diff
@@ -64,46 +68,109 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
+            }
+            catch (ResourceLimitExceededException ex) {
+                throw new TaskExecutionException(ex);
+            }
 
-                boolean isDone = attempt.getStateFlags().isDone();
-                if (isDone) {
-                    if (!ignoreFailure && !attempt.getStateFlags().isSuccess()) {
+            if (sessionAttemptConflictException.isPresent()) {
+                StoredSessionAttempt conflictedAttempt = sessionAttemptConflictException.get().getConflictedSession();
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:132]: How about moving this block to `catch (SessionAttemptConflictException ex) { ... }` block? Maybe we can remove the Optional?

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 4]
```diff
@@ -64,46 +68,109 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
-                    // TODO use exponential-backoff to calculate retry interval
-                    throw TaskExecutionException.ofNextPolling(1, ConfigElement.copyOf(request.getLastStateParams()));
+                    // Wait for finish running attempt
+                    throw nextPolling(request.getLastStateParams().deepCopy());
                 }
             }
-            catch (ResourceNotFoundException ex) {
-                throw new ConfigException(ex);
+            else if (attempt.isPresent()) { // startSession succeeded and created new attempt
+                throw nextPolling(request.getLastStateParams().deepCopy().set("require_kicked", true));
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:157]: How about moving this line to right after calling `callback.startSession()`  in the above `try { ... }` block? Maybe we can remove the Optional?

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 3 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:132]: Of course, we can do.  My intention of this way is that firstly get result of startSession() including exception (especially SessionAttemptConflictException is not error), then the logic goes ahead based on the result.
if we move the logic in `catch (SessionAttemptConflictException ex) { ... }`, catch {} become large.  
In this case it is better to define another method and call it.
What is your thought?

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 4 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:157]: My intention is same as above. But this code is small. 
So I agree with move to `try{..}`.

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 2 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:100]: Do you mean, if the attempt which is kicked by require> failed,  what happen?
In this case the task will fail as usual.

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 1 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:125]: Thanks! I will fix them.

Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java]: Fixed it. Thanks!

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 3 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java]: Based on discussion in person, I fixed it.
As result, the code became simple. Thank you for your advice! 🙂

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 4 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java]: Fixed it. Thanks!

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 5]
```diff
@@ -65,43 +68,149 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
+
             String workflowName = config.get("_command", String.class);
-            int projectId = config.get("project_id", int.class);
             Instant instant = config.get("session_time", Instant.class);
             boolean ignoreFailure = config.get("ignore_failure", boolean.class, false);
-            Optional<String> retryAttemptName = config.getOptional("retry_attempt_name", String.class);
+            OptionRerunOn rerunOn = OptionRerunOn.of(config.get("rerun_on", String.class, "none"));
+
+            //retry_attempt_name is set from local config only.
+            Optional<String> retryAttemptName = localConfig.getOptional("retry_attempt_name", String.class);
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:80]: I'm wondering if this scope reduction of `retry_attempt_name` can affect existing workflows...

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 6]
```diff
@@ -65,43 +68,149 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
+        }
+    }
+
+    enum OptionRerunOn
+    {
+        NONE("none"),
+        FAILED( "failed"),
+        ALL ("all");
+
+        private final String name;
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:201]: This isn't needed?

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 7]
```diff
@@ -65,43 +68,149 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
             }
         }
+
+        private TaskResult processAttempt(StoredSessionAttempt attempt, Config lastStateParams, OptionRerunOn rerunOn, boolean ignoreFailure)
+        {
+            if (attempt.getStateFlags().isDone()) {
+                // A flag to distinguish whether the attempt is kicked by require> or previous attempt.
+                boolean requireKicked = lastStateParams.get("require_kicked", boolean.class, false);
+                if (!requireKicked &&  (
+                        rerunOn == OptionRerunOn.ALL ||
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:137]: When OptionRerunOn.ALL is set, the workflow should be executed regardless it was executed via `require` operator or not, right? If so, `rerunOn == OptionRerunOn.ALL` should not depend on `requireKicked`?

Comment (komamitsu) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:137]: Ah, `require_kicked` is marked on the current request not a previous attempt...

Comment (komamitsu) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:137]: Ah, got it. I had a wrong idea. It's for `Check attempt kicker` on the diagram.

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 8]
```diff
@@ -65,43 +68,149 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
+            }
+        }
+
+        private TaskExecutionException nextPolling(Config stateParams)
+        {
+            int iteration = stateParams.get("retry", int.class, 0);
+            stateParams.set("retry", iteration+1);
+            int interval = (int) Math.min(1 * Math.pow(2, iteration), MAX_TASK_RETRY_INTERVAL);
+
+            return TaskExecutionException.ofNextPolling( interval, ConfigElement.copyOf(stateParams));
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:163]: A space on the left of `interval`

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 9]
```diff
@@ -65,43 +68,149 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
+            else {
+                // Wait for finish running attempt
+                throw nextPolling(request.getLastStateParams().deepCopy());
+            }
+        }
+
+        private TaskExecutionException nextPolling(Config stateParams)
+        {
+            int iteration = stateParams.get("retry", int.class, 0);
+            stateParams.set("retry", iteration+1);
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:160]: Spaces on the both sides of `+` would be better?

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java [hunk 10]
```diff
@@ -65,43 +68,149 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
...
+                    // ignore_failure is false and the attempt is in error state. Make this operator failed.
+                    throw new TaskExecutionException(String.format(ENGLISH,
+                            "Dependent workflow failed. Session id: %d, attempt id: %d",
+                            attempt.getSessionId(), attempt.getId()));
+                }
+                return TaskResult.empty(cf);
+            }
+            else {
+                // Wait for finish running attempt
+                throw nextPolling(request.getLastStateParams().deepCopy());
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:153]: Maybe duplicated deep copy?

File: digdag-docs/src/operators/require.md [hunk 11]
```diff
@@ -71,3 +88,8 @@
   ignore_failure: true
   ```
 
+  require> evaluates *ignore_failure* at last of its process. If rerun_on is set and require> run new attempt, the result of new attempt is checked.
+
+## Notes
+- require> has been changed to ignore inherited *retry_attempt_name* parameter. 
+  `digdag retry ` command generates unique retry_attempt_name to run, but it is not passed to require>.
```
Comment (komamitsu) [inline digdag-docs/src/operators/require.md:95]: Unnecessary space on the right of `digdag retry`

File: digdag-docs/src/operators/require.md [hunk 12]
```diff
@@ -71,3 +88,8 @@
   ignore_failure: true
   ```
 
+  require> evaluates *ignore_failure* at last of its process. If rerun_on is set and require> run new attempt, the result of new attempt is checked.
+
+## Notes
+- require> has been changed to ignore inherited *retry_attempt_name* parameter.
```
Comment (komamitsu) [inline digdag-docs/src/operators/require.md:94]: Adding `since v0.10` or something may be helpful for users, I think.

Comment (yoyama) [reply digdag-docs/src/operators/require.md:94]: Good point. But I am planning back port to 0.9.xx.

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 6 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:201]: Ah, correct. Currently not used. Will remove it. Thanks!

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 5 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:80]: Yes. There is possibility as discussion in person.
Let me discuss again in a meeting.

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 10 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:153]: May be not duplicated, but I should use lastStateParam argument in my understand.

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 5 above)
Comment (komamitsu) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java:80]: Just noticed you've already mentioned it in the document :)

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 8 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java]: Thanks! Fixed it.

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 9 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java]: Thanks! Fixed it.

File: digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java (hunk 10 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java]: I updated it.

File: digdag-docs/src/operators/require.md (hunk 11 above)
Comment (yoyama) [reply digdag-docs/src/operators/require.md]: Fixed it. Thanks!

File: digdag-tests/src/test/java/acceptance/RequireIT.java [hunk 13]
```diff
@@ -185,6 +203,159 @@ public void testIgnoreProjectIdParam()
...
+                "-e", server.endpoint());
+        assertThat(pushParentStatus.errUtf8(), pushParentStatus.code(), is(0));
+
+        CommandStatus startStatus = main("start",
+                "-c", config.toString(),
+                "-e", server.endpoint(),
+                "parent_another", "parent_another_project",
+                "--session", "now"
+        );
+        assertThat(startStatus.errUtf8(), startStatus.code(), is(0));
```
Comment (komamitsu) [inline digdag-tests/src/test/java/acceptance/RequireIT.java:250]: Just posted a comment https://github.com/treasure-data/digdag/pull/1405/files#r435112975

File: digdag-tests/src/test/java/acceptance/RequireIT.java [hunk 14]
```diff
@@ -185,6 +203,159 @@ public void testIgnoreProjectIdParam()
...
+            RestSessionAttemptCollection childAttempts = testRerunOnParam(sessionTime, "failed", false);
+            assertThat("Number of child's attempt must be one. (== require> kick)", childAttempts.getAttempts().size(), is(1));
+        }
+    }
+
+    private RestSessionAttemptCollection testRerunOnParam(String sessionTime, String rerunOn, boolean previousChildRunFail) throws InterruptedException
+    {
+        //Start a child
+        String childFailParam = previousChildRunFail ? "yes" : "no";
+        CommandStatus childStatus = startAndWait(true,"require", "child", "--session", sessionTime,
```
Comment (komamitsu) [inline digdag-tests/src/test/java/acceptance/RequireIT.java:310]: [minor] Space

File: digdag-tests/src/test/java/acceptance/RequireIT.java [hunk 15]
```diff
@@ -185,6 +203,159 @@ public void testIgnoreProjectIdParam()
...
+                "-p", "child_fail=" + childFailParam);
+        assertThat(isAttemptSuccess(childStatus), not(previousChildRunFail));
+
+        // Start a parent with same session time.
+        CommandStatus rerunOnNoneStatus = startAndWait(false,"require", "parent", "--session", sessionTime,
```
Comment (komamitsu) [inline digdag-tests/src/test/java/acceptance/RequireIT.java:315]: Same as above

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 14 above)
Comment (komamitsu) [inline digdag-tests/src/test/java/acceptance/RequireIT.java:310]: I'm wondering why this method call always passes `ignoreAttemptFailure=true` (sorry, a bit too difficult for me 😅.)

Could I ask you the followings?
1. Add a comment for each case about why target attempt is expected to fail
2. If possible, make this method call pass `attemptShouldFail` or something deterministic instead of ` ignoreAttemptFailure`

File: digdag-tests/src/test/java/acceptance/RequireIT.java [hunk 16]
```diff
@@ -185,6 +203,159 @@ public void testIgnoreProjectIdParam()
...
+                success = true;
+                break;
+            }
+            else if (attemptsStatus.outUtf8().contains("status: error")) {
+                break;
+            }
+            Thread.sleep(1000);
+        }
+        if (!ignoreAttemptFailure) {
+            assertThat(success, is(true));
```
Comment (komamitsu) [inline digdag-tests/src/test/java/acceptance/RequireIT.java:353]: Oh, the same check is done above like `assertThat(isAttemptSuccess(childStatus), not(previousChildRunFail))`?
So we can remove this to make things simpler?

File: digdag-tests/src/test/java/acceptance/RequireIT.java [hunk 17]
```diff
@@ -185,6 +203,159 @@ public void testIgnoreProjectIdParam()
...
+            throws Exception
+    {
+        // Create new project
+        CommandStatus initStatus = main("init",
+                "-c", config.toString(),
+                projectDir.toString());
+        assertThat(initStatus.errUtf8(), initStatus.code(), is(0));
+
+        Path childOutFile = projectDir.resolve("child.out").toAbsolutePath().normalize();
+        prepareForChildWF(childOutFile);
```
Comment (komamitsu) [inline digdag-tests/src/test/java/acceptance/RequireIT.java:264]: Is this line needed?

File: digdag-tests/src/test/java/acceptance/RequireIT.java [hunk 18]
```diff
@@ -185,6 +203,159 @@ public void testIgnoreProjectIdParam()
...
+                "require",
+                "-c", config.toString(),
+                "-e", server.endpoint(),
+                "-r", "4711");
+        assertThat(pushStatus.errUtf8(), pushStatus.code(), is(0));
+
+        // test for rerun_on: none
+        {
+            String sessionTime = "2020-05-28 12:34:01";
+            RestSessionAttemptCollection childAttempts = testRerunOnParam(sessionTime, "none", false);
```
Comment (komamitsu) [inline digdag-tests/src/test/java/acceptance/RequireIT.java:280]: How about adding a test case of the combination of `none` and true?

File: digdag-tests/src/test/java/acceptance/RequireIT.java [hunk 19]
```diff
@@ -185,6 +203,159 @@ public void testIgnoreProjectIdParam()
...
+            assertThat("Number of child's attempt must be one. (== require> skip the call)", childAttempts.getAttempts().size(), is(1));
+        }
+
+        // test for rerun_on: all. require> kick child always.
+        {
+            String sessionTime = "2020-05-28 12:34:11";
+            RestSessionAttemptCollection childAttempts = testRerunOnParam(sessionTime, "all", false);
```
Comment (komamitsu) [inline digdag-tests/src/test/java/acceptance/RequireIT.java:287]: How about adding a test case of the combination of `all` and true?

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 14 above)
Comment (komamitsu) [reply digdag-tests/src/test/java/acceptance/RequireIT.java:310]: After having read all codes in `testRerunOnParam()` I understand all things. So probably you can ignore the above comment from me.

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 17 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java:264]: Thanks. I will remove 263 and 264.

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 16 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java:353]: Thanks for your pointing out. I will remove them and argument ignoreAttemptFailure.

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 14 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java:310]: I will add some additional comments.

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 13 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java:251]: I fixed it in #1405 and if it become OK, rebase to this PR.

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 19 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java]: I added it. Thanks!

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 18 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java]: Thanks, I added it. Now all conditions are covered.

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 17 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java]: Fixed it, thanks.

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 16 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java]: Fixed it and refactoring. Thanks for your pointing out.

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 14 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java]: Done. Thanks!

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 15 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java]: Fixed it. (Removed the param because of refactoring)

File: digdag-tests/src/test/java/acceptance/RequireIT.java (hunk 14 above)
Comment (yoyama) [reply digdag-tests/src/test/java/acceptance/RequireIT.java]: Fixed it. (Removed the param because of refactoring)

File: digdag-tests/src/test/java/acceptance/RequireIT.java [hunk 20]
```diff
@@ -185,6 +203,176 @@ public void testIgnoreProjectIdParam()
...
+    /**
+     * Test for rerun_on
+     *
+     * @param sessionTime
+     * @param rerunOn              parameter for rerun_on:  "none" or "fail" or "all"
+     * @param previousChildRunFail if true, first child workflow failed.
+     * @return
+     * @throws InterruptedException
+     */
+    private RestSessionAttemptCollection testRerunOnParam(String sessionTime, String rerunOn, boolean previousChildRunFail, boolean expectParentSucceed) throws InterruptedException
```
Comment (komamitsu) [inline digdag-tests/src/test/java/acceptance/RequireIT.java:328]: 💯

File: digdag-tests/src/test/resources/acceptance/require/parent_rerun_on.dig [hunk 21]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (komamitsu) [inline digdag-tests/src/test/resources/acceptance/require/parent_rerun_on.dig:1]: Empty line?

**Comment by yoyama** [comment]
@komamitsu  Thank you so much for your long reviews.
It is very helpful. After fix of #1405 is approved, I will rebase it then merge this.
